src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs

```diff
@@ -0,0 +1,289 @@
+                char ch = (char)i;
+                Debug.Assert((allowedCharacters.IsCharacterAllowed(ch) ? 1 : 0) == AllowList[ch]);
+                Debug.Assert(allowedCharacters.IsCharacterAllowed(ch) == !NeedsEscaping(ch));
+            }
+            for (int i = AllowList.Length; i <= char.MaxValue; i++)
+            {
+                char ch = (char)i;
+                Debug.Assert(!allowedCharacters.IsCharacterAllowed(ch));
+                Debug.Assert(NeedsEscaping(ch));
+            }
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public override bool WillEncode(int unicodeScalar)
+        {
+            if (UnicodeHelpers.IsSupplementaryCodePoint(unicodeScalar))
+            {
+                return true;
+            }
+
+            Debug.Assert(unicodeScalar >= char.MinValue && unicodeScalar <= char.MaxValue);
+
+            return NeedsEscaping((char)unicodeScalar);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public override unsafe int FindFirstCharacterToEncode(char* text, int textLength)
+        {
+            if (text == null)
+            {
+                throw new ArgumentNullException(nameof(text));
+            }
+
+            int idx = 0;
+
+#if BUILDING_INBOX_LIBRARY
+            if (Sse2.IsSupported)
+            {
+                short* startingAddress = (short*)text;
+                while (textLength - 8 >= idx)
+                {
+                    Debug.Assert(startingAddress >= text && startingAddress <= (text + textLength - 8));
+
+                    // Load the next 8 characters.
+                    Vector128<short> sourceValue = Sse2.LoadVector128(startingAddress);
+
+                    // Check if any of the 8 characters need to be escaped.
+                    Vector128<short> mask = Sse2Helper.CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin(sourceValue);
+
+                    int index = Sse2.MoveMask(mask.AsByte());
+                    // If index == 0, that means none of the 8 characters needed to be escaped.
+                    // TrailingZeroCount is relatively expensive, avoid it if possible.
+                    if (index != 0)
+                    {
+                        // Found at least one character that needs to be escaped, figure out the index of
+                        // the first one found that needed to be escaped within the 8 characters.
+                        Debug.Assert(index > 0 && index <= 65_535);
+                        int tzc = BitOperations.TrailingZeroCount(index);
+                        Debug.Assert(tzc % 2 == 0 && tzc >= 0 && tzc <= 16);
+                        idx += tzc >> 1;
+                        goto Return;
+                    }
+                    idx += 8;
+                    startingAddress += 8;
+                }
+
+                // Process the remaining characters.
+                Debug.Assert(textLength - idx < 8);
+            }
+#endif
```

Comment:
This is the crux of the optimization (applied to both utf-16 and utf-8 overloads for the different `JavaScriptEncoders`).